From bc3c8d3288a78ffb723173dbdfb2d60c2a4e4deb Mon Sep 17 00:00:00 2001
From: Aniruddha C <aniruddhac@nvidia.com>
Date: Mon, 8 Oct 2012 18:16:53 +0530
Subject: [PATCH 1/6] [SF Streaming] Abort if timestamp rollover happens

If server starts sending the data from 0 secs, then
video will hang since there is no seek called and hence
buffers at the ports are not flushed. Hence, abort if
timestamp rollover happens.

Bug 1049841

Change-Id: I5153e670bcb748e1c8dba9de3d6593655b88b80c
---
 media/libstagefright/rtsp/MyHandler.h |   10 ++++++++++
 1 files changed, 10 insertions(+), 0 deletions(-)

diff --git a/media/libstagefright/rtsp/MyHandler.h b/media/libstagefright/rtsp/MyHandler.h
index f063b54..4bca34e 100644
--- a/media/libstagefright/rtsp/MyHandler.h
+++ b/media/libstagefright/rtsp/MyHandler.h
@@ -125,6 +125,7 @@ struct MyHandler : public AHandler {
           mSetupTracksSuccessful(false),
           mSeekPending(false),
           mFirstAccessUnit(true),
+          mFirstPlayRespParsed(false),
           mAllTracksHaveTime(false),
           mNTPAnchorUs(-1),
           mMediaAnchorUs(-1),
@@ -664,6 +665,7 @@ struct MyHandler : public AHandler {
                         result = UNKNOWN_ERROR;
                     } else {
                         parsePlayResponse(response);
+                        mFirstPlayRespParsed = true;
 
                         sp<AMessage> timeout = new AMessage('tiou', id());
                         timeout->post(kStartupTimeoutUs);
@@ -749,6 +751,7 @@ struct MyHandler : public AHandler {
                 mSetupTracksSuccessful = false;
                 mSeekPending = false;
                 mFirstAccessUnit = true;
+                mFirstPlayRespParsed = false;
                 mAllTracksHaveTime = false;
                 mNTPAnchorUs = -1;
                 mMediaAnchorUs = -1;
@@ -1294,6 +1297,12 @@ struct MyHandler : public AHandler {
             return;
         }
 
+        if (!mSeekPending && mFirstPlayRespParsed && (npt1 == 0.0)) {
+            ALOGI("Rollover occured. Abort now...");
+            (new AMessage('abor', id()))->post();
+            return;
+        }
+
         i = response->mHeaders.indexOfKey("rtp-info");
         CHECK_GE(i, 0);
 
@@ -1418,6 +1427,7 @@ private:
     bool mSeekPending;
     List<int64_t> mSeekPoints;
     bool mFirstAccessUnit;
+    bool mFirstPlayRespParsed;
 
     bool mAllTracksHaveTime;
     int64_t mNTPAnchorUs;
-- 
1.7.1

